test/pylib/pool.py

```diff
@@ -81,6 +91,22 @@ async def steal(self) -> None:
             self.total -= 1
             self.cond.notify()
 
+    async def steal_and_get(self, *args, **kwargs) -> T:
+        """Atomically take ownership of a previously borrowed object
+           which frees up space in the pool, and borrow a new object.
+
+           *args and **kwargs are used as in `get`.
+        """
+        async with self.cond:
+            if self.pool:
+                self.total -= 1
+                return self.pool.pop()
+
+            # Need to construct a new object.
+            # The space for this object is already accounted for in self.total.
+
+        return await self._build_and_get(*args, **kwargs)
```

Comment:
So here we're building a new cluster while the previous one is "stolen", so we're temporarily going beyond the pool size. This is sad.

Reply:
Come to think of it, I am more and more convinced that we should not let outside users destroy clusters created by the pool, i.e. "steal" them. We should always be returning the cluster to the pool, if only to destroy it and build a new one for us. The pool then can do it atomically, with proper accounting of the waiters and the pool limits. 

Please consider killing steal() and replacing steal_and_get with something like replace_dirty_cluster_with_clean(dirty_cluster: Cluster) -> Cluster.

Reply:
> So here we're building a new cluster while the previous one is "stolen", so we're temporarily going beyond the pool size. This is sad.

Depends on the user. Whether they destroy the object before or after stealing it.
I modified the code so we stop the cluster before steal_and_get, so we're not going beyond pool size.